hw/xbox/nv2a/pgraph.c

```
@@ -7176,15 +7219,269 @@ static bool texture_cache_entry_compare(Lru *lru, LruNode *node, void *key)
     return memcmp(&tnode->key, key, sizeof(TextureKey));
 }
 
-/* hash and equality for shader cache hash table */
-static guint shader_hash(gconstpointer key)
+static void shader_cache_entry_init(Lru *lru, LruNode *node, void *state)
+{
+    ShaderLruNode *snode = container_of(node, ShaderLruNode, node);
+    memcpy(&snode->state, state, sizeof(ShaderState));
+    snode->cached = false;
+    snode->binding = NULL;
+    snode->program = NULL;
+}
+
+static void shader_cache_entry_post_evict(Lru *lru, LruNode *node)
 {
-    return fast_hash((const uint8_t *)key, sizeof(ShaderState));
+    ShaderLruNode *snode = container_of(node, ShaderLruNode, node);
+    
+    if (snode->cached) {
+        qemu_thread_join(&snode->save_thread);
+    }
+    
+    if (snode->binding) {
+        glDeleteProgram(snode->binding->gl_program);
+        g_free(snode->binding);
+    }
+
+    if (snode->program) {
+        g_free(snode->program);
+    }
+}
+
+static bool shader_cache_entry_compare(Lru *lru, LruNode *node, void *key)
+{
+    ShaderLruNode *snode = container_of(node, ShaderLruNode, node);
+    return memcmp(&snode->state, key, sizeof(ShaderState));
+}
+
+static char *shader_get_lru_cache_path(void)
+{
+    char *cfg_dir = xemu_get_config_dir_path();
+    size_t path_len = strlen(cfg_dir) + 19;
+    char *path = g_malloc(path_len);
+    snprintf(path, path_len, "%s/shader_cache_list", cfg_dir);
+    g_free(cfg_dir);
+    return path;
+}
+
+static bool shader_load_from_memory(ShaderLruNode *snode)
+{
+    assert(glGetError() == GL_NO_ERROR);
+
+    if (!snode->program) {
+        return false;
+    }
+
+    GLuint gl_program = glCreateProgram();
+    glProgramBinary(gl_program, snode->program_format, snode->program, snode->program_size);
+    if (glGetError() != GL_NO_ERROR) {
```

This seems potentially trace-worthy (w/ the actual error preserved).

---

Making sure I understand the flow: the only way we'd anticipate hitting those would be if the shader loaded from disk has become invalid (either due to it being corrupted or the driver being updated). I think if you switch to doing explicit validation on the reload case it should no longer be possible for this to be hit?

If that's the case, then should we just assert instead of handling it silently-but-gracefully (and potentially hiding some failure mode)?

---

That's correct. This path is happening on load, and I don't know if there's another way to check if the binary format is supported beforehand (or if it's worth doing, since the GL implementation will do it here as well).